core/src/test/scala/unit/kafka/controller/ControllerEventManagerTest.scala

```diff
@@ -0,0 +1,94 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kafka.controller
+
+import java.util.concurrent.atomic.AtomicInteger
+
+import com.yammer.metrics.Metrics
+import com.yammer.metrics.core.Timer
+import kafka.utils.TestUtils
+import org.easymock.{EasyMock, IAnswer}
+import org.junit.Test
+import org.junit.Assert.{assertEquals, fail}
+
+import scala.collection.JavaConverters._
+
+class ControllerEventManagerTest {
+
+  private var controllerEventManager: ControllerEventManager = _
+
+  def tearDown(): Unit = {
+    if (controllerEventManager != null)
+      controllerEventManager.close()
+  }
+
+  @Test
+  def testSuccessfulEvent(): Unit = {
+    check("kafka.controller:type=ControllerStats,name=AutoLeaderBalanceRateAndTimeMs", ControllerState.AutoLeaderBalance,
+      () => Unit)
+  }
+
+  @Test
+  def testEventThatThrowsException(): Unit = {
+    check("kafka.controller:type=ControllerStats,name=LeaderElectionRateAndTimeMs", ControllerState.BrokerChange,
+      () => throw new NullPointerException)
+  }
+
+  private def check(metricName: String, controllerState: ControllerState, process: () => Unit): Unit = {
+    val controllerStats = new ControllerStats
+    val eventProcessedListenerCount = new AtomicInteger
+    controllerEventManager = new ControllerEventManager(controllerStats.rateAndTimeMetrics,
+      _ => eventProcessedListenerCount.incrementAndGet)
+    controllerEventManager.start()
+
+    val initialTimerCount = timer(metricName).count
+
+    // `ControllerEvent` is sealed so we use EasyMock to create a subclass
+    val eventMock = EasyMock.createMock(classOf[ControllerEvent])
+    EasyMock.expect(eventMock.state).andReturn(controllerState)
+
+    // Continue processing until we have checked the controller state
+    @volatile var processing = true
+    EasyMock.expect(eventMock.process()).andAnswer(new IAnswer[Unit]() {
+      def answer(): Unit = {
+        while (processing)
+          Thread.`yield`()
```

Comment:
That's a good suggestion. Both are correct (even if `Thread.yield` is a no-op, it won't matter here), but `CountDownLatch` is more concise and shows intent better. Updated the PR.